src/desktop/credentials/DesktopNativeCredentialsFacade.ts

```diff
@@ -1,35 +1,136 @@
+/**
+ *
+ */
 export class DesktopNativeCredentialsFacade implements NativeCredentialsFacade {
-	private readonly _desktopKeyStoreFacade: DesktopKeyStoreFacade
-	private readonly _crypto: DesktopNativeCryptoFacade
+	/**
+	 * @param desktopKeyStoreFacade
+	 * @param crypto
+	 * @param argon2idFacade
+	 * @param lang
+	 * @param conf
+	 * @param getCurrentCommonNativeFacade a "factory" that returns the commonNativeFacade for the window that would be most suited to serve a given request
+	 */
+	constructor(
+		private readonly desktopKeyStoreFacade: DesktopKeyStoreFacade,
+		private readonly crypto: DesktopNativeCryptoFacade,
+		private readonly argon2idFacade: Argon2idFacade,
+		private readonly lang: LanguageViewModel,
+		private readonly conf: DesktopConfig,
+		private readonly getCurrentCommonNativeFacade: () => Promise<CommonNativeFacade>,
+	) {}
 
-	constructor(keyStoreFacade: DesktopKeyStoreFacade, crypto: DesktopNativeCryptoFacade) {
-		this._desktopKeyStoreFacade = keyStoreFacade
-		this._crypto = crypto
+	async decryptUsingKeychain(data: Uint8Array, encryptionMode: DesktopCredentialsMode): Promise<Uint8Array> {
+		// making extra sure that the mode is the right one since this comes over IPC
+		this.assertSupportedEncryptionMode(encryptionMode)
+		data = await this.maybeDecryptUsingAppPin(data, encryptionMode)
+		const credentialsKey = await this.desktopKeyStoreFacade.getCredentialsKey()
+		data = this.crypto.aes256DecryptKey(credentialsKey, data)
+		return data
 	}
 
-	async decryptUsingKeychain(data: Uint8Array, encryptionMode: CredentialEncryptionMode.DEVICE_LOCK): Promise<Uint8Array> {
+	async encryptUsingKeychain(data: Uint8Array, encryptionMode: DesktopCredentialsMode): Promise<Uint8Array> {
 		// making extra sure that the mode is the right one since this comes over IPC
-		assert(encryptionMode === CredentialEncryptionMode.DEVICE_LOCK, "should not use unsupported encryption mode")
-		const key = await this._desktopKeyStoreFacade.getCredentialsKey()
-		const decryptedData = this._crypto.aes256DecryptKey(key, data)
-		return Promise.resolve(decryptedData)
+		this.assertSupportedEncryptionMode(encryptionMode)
+		const credentialsKey = await this.desktopKeyStoreFacade.getCredentialsKey()
+		data = this.crypto.aes256EncryptKey(credentialsKey, data)
+		data = await this.maybeEncryptUsingAppPin(data, encryptionMode)
+		return data
 	}
 
-	async encryptUsingKeychain(data: Uint8Array, encryptionMode: CredentialEncryptionMode.DEVICE_LOCK): Promise<Uint8Array> {
-		// making extra sure that the mode is the right one since this comes over IPC
-		assert(encryptionMode === CredentialEncryptionMode.DEVICE_LOCK, "should not use unsupported encryption mode")
-		const key = await this._desktopKeyStoreFacade.getCredentialsKey()
-		const encryptedData = this._crypto.aes256EncryptKey(key, data)
-		return Promise.resolve(encryptedData)
+	private async maybeDecryptUsingAppPin(data: Uint8Array, encryptionMode: DesktopCredentialsMode): Promise<Uint8Array> {
```

Comment:
I find `maybe` part of the name a bit confusing, like it might fail. I would probably only have `decryptUsingPin()` but if you keep this one how about `decryptingWithPinIfNeeded()`?